core/src/main/java/hivemall/evaluation/FMeasureUDAF.java

```diff
@@ -18,118 +18,387 @@
+            long totalActual = PrimitiveObjectInspectorFactory.writableLongObjectInspector.get(totalActualObj);
+            long totalPredicted = PrimitiveObjectInspectorFactory.writableLongObjectInspector.get(totalPredictedObj);
+            double beta = PrimitiveObjectInspectorFactory.writableDoubleObjectInspector.get(betaObj);
+            String average = PrimitiveObjectInspectorFactory.writableStringObjectInspector.getPrimitiveJavaObject(averageObj);
+
+            FMeasureAggregationBuffer myAggr = (FMeasureAggregationBuffer) agg;
+            myAggr.merge(tp, totalActual, totalPredicted, beta, average);
+        }
+
+        @Override
+        public DoubleWritable terminate(@SuppressWarnings("deprecation") AggregationBuffer agg)
+                throws HiveException {
+            FMeasureAggregationBuffer myAggr = (FMeasureAggregationBuffer) agg;
+            double result = myAggr.get();
+            return new DoubleWritable(result);
+        }
+    }
+
+    public static class FMeasureAggregationBuffer extends
+            GenericUDAFEvaluator.AbstractAggregationBuffer {
+        long tp;
+        /** tp + fn */
+        long totalActual;
+        /** tp + fp */
+        long totalPredicted;
+        double beta;
+        String average;
+
+        public FMeasureAggregationBuffer() {
+            super();
         }
 
-        /**
-         * @return 2 * precision * recall / (precision + recall)
-         */
-        private static double f1Score(final PartialResult partial) {
-            double precision = precision(partial);
-            double recall = recall(partial);
-            double divisor = precision + recall;
+        void setOptions(double beta, String average) {
+            this.beta = beta;
+            this.average = average;
+        }
+
+        void reset() {
+            this.tp = 0L;
+            this.totalActual = 0L;
+            this.totalPredicted = 0L;
+        }
+
+        void merge(long o_tp, long o_actual, long o_predicted, double beta, String average) {
+            tp += o_tp;
+            totalActual += o_actual;
+            totalPredicted += o_predicted;
+            this.beta = beta;
+            this.average = average;
+        }
+
+        double get() {
+            double squareBeta = Math.pow(beta, 2.d);
+            double divisor;
+            double numerator;
+
+            if (average.equals("micro")) {
+                divisor = denom(tp, totalActual, totalPredicted, squareBeta);
+                numerator = (1.d + squareBeta) * tp;
+            } else if (average.equals("binary")) {
+                double precision = precision(tp, totalPredicted);
+                double recall = recall(tp, totalActual);
+                divisor = squareBeta * precision + recall;
+                numerator = (1.d + squareBeta) * precision * recall;
+            } else {
+                throw new UnsupportedOperationException();
```

Comment:
Yes, you are right. Currently, `average` value is only `binary` or `micro`.
So, I remove this `else` statement and replace `else if ` to `else`.